src/app/files.js

```diff
@@ -107,6 +107,47 @@ function Files (storage) {
     return files
   }
 
+  //
+  // Tree model for files
+  // {
+  //   'a': { }, // empty directory 'a'
+  //   'b': {
+  //     'c': {}, // empty directory 'b/c'
+  //     'd': { '/readonly': true, '/content': 'Hello World' } // files 'b/c/d'
+  //     'e': { '/readonly': false, '/path': 'b/c/d' } // symlink to 'b/c/d'
```

Comment:
I was wondering if a simple straight forward nested object would maybe be enough.  
If it really matters whether the structure can support all or lots of stuff normal file systems support, then maybe it should follow some kind of "standard".  

I was thinking a bit ... and in case i'll use the result of `files.listAsTree()` i have to refactor the `file-explorer` anyway... and was checking `npm` a bit and found:
* **https://www.npmjs.com/package/fs-file-tree** (my favourite so far regarding "inspiration")
* https://www.npmjs.com/package/files-tree
* https://www.npmjs.com/package/treeify-paths
* https://www.npmjs.com/package/fsify
* https://www.npmjs.com/package/dir-to-json
* ... maybe there are more examples ...

---

So what about leaning on `require('fs').stats` method and do something like:

```js
// representation of a single node
// (given that "" (empty string) as a property key is valid json)
var exampleNode = {
  "name": {
    "stats": {
      "type": 'directory', // one of: ['file', 'directory', 'blockdevice' , 'characterdevice', 'symlink', 'FIFO', 'socket']
      // ... and a list of all the properties supported by https://nodejs.org/api/fs.html#fs_class_fs_stats
    },
    "": { /* .... */ }
  }
}
```
assuming the "meta data" will be contained in the [.stats property]( https://nodejs.org/api/fs.html#fs_class_fs_stats),  
an example file system object could look like the following:

```js
var fs = {
  "a" : {
    "stats": {
      "type": "directory"
      // ...
    },
    "": { }
  },
  "b": {
    "stats": {
      "type": "directory"
      // ...
    },
    "": {
      "c": {
        "stats": {
          "type": "directory"
          // ...
        },
        "": { }
      },
      "d": {
        "stats": {
          "type": "file",
          "mode": "r"
          // ...
        },
        "": "Hello World"
      },
      "e": {
        "stats": {
          "type": "symlink",
          "mode": "rw"
          // ...
        },
        "": ["b", "f"]
      },
      "f": {
        "stats": {
          "type": "file",
          "mode": "rwx"
          // ...
        },
        "": "...maybe javascript content or base64 encoded stuff or whatever ..."
        // "": function () { console.log('executable stuff') }
      }
    }
  }
}

function get (fs, path, query) {
  query = query ? 'stats' : ''
  var node = { '': path }
  do {
    node = node[''].reduce((f, p, i, path) => (i !== path.length - 1) ? f[p][''] : f[p], fs)
  } while (node.stats.type === 'symlink')
  return node[query]
}

// usage
function getContent (path) { return get(fs, path) }
function getStats (path) { return get(fs, path, 'stats') }

var path = 'b/e'.split('/')

var stats = getStats(path)
var content = getContent(path)

```
If an executable node would ACTUALLY be content wrapped in a callable function, that would be super awesome i think :P ...would be so cool to use. And if you set random garbage to be executable, it would just throw, but that's fine, because making random garbage on a real file system executable also does not work.